### PR TITLE
kmsv2: implement expire cache with clock 

### LINKED ISSUE #111920: [KMSv2] Use expiring cache
https://github.com/kubernetes/kubernetes/blob/d581cc90adba6c84919738841fe3e07302d53e33/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go#L95-L99

/sig auth

### PR DESCRIPTION
**Pull request by aramase**
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
- Use expiring cache for KMS v2
  - The default TTL for cache entry is set to 1h

#### Which issue(s) this PR fixes:

Fixes #111920

#### Does this PR introduce a user-facing change?

```release-note
cacheSize field in EncryptionConfiguration is not supported for KMSv2 provider
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3299-kms-v2-improvements
```

### CODE REVIEW

**Comment by aramase** [comment]
/assign @enj
/hold

(work in progress)

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:200]: I would move both the base64 and cache record bits into the cache and let the caller pass in the encryption key bytes and the transformer

```go
set(key []bytes, transformer value.Transformer)
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:175]: `testEnvelopeCacheSize` is probably a const that can be deleted.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:68]: Is this method used?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:109]: I am trying to think of how we can more rigorously test the new cache type...

**Comment by aramase** [comment]
/cc @mikedanese 

@mikedanese Could you take a look at the PR when you get a chance?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go:36]: Reading through the code I actually think we want:

```suggestion
		cache: utilcache.NewLRUExpireCacheWithClock(clock),
```

This cache supports both TTL based expiration while limiting the max size of the cache.

Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go:36]: `NewLRUExpireCacheWithClock` seems to be missing a GC loop but that should be easy enough to add.

**Review by enj — CHANGES_REQUESTED** [review]
This is close.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go
Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go:36]: sounds good! opened https://github.com/kubernetes/kubernetes/issues/115174

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go
Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:175]: good catch! I had removed those tests because we were using just the expiring cache before. Adding the test back now.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go
Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go]: done!

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:53]:
```suggestion
	// cacheMaxSize is the maximum size of the cache.
	// This is meant to limit the size of the cache to around ~100 MB:
	// Each EDEK (the cache key) has a max size of 1 KB
	// Each DEK (the cache value) has a size of 32 bytes
	cacheMaxSize = 100 * 1024
```

**Review by enj — COMMENTED** [review]
Just one nit.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go
Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go]: thanks, done!

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:181]: Please create an issue for this (can be the same issue that tracks the new cache implementation work).  And for adding `TestEnvelopeCacheLimit` back (if it makes sense for the new cache).

**Review by enj — APPROVED** [review]
/lgtm
/approve

/assign @liggitt 
(for review on API defaulting and validation changes for KMS v2)

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go
Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:181]: opened https://github.com/kubernetes/kubernetes/issues/115275

**Comment by aramase** [comment]
@enj had to push an update removing `cacheSize` from one of the new added integration tests, need another `lgtm` 😄

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1/defaults.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1/defaults.go:46]: update the godoc on the CacheSize field to indicate the field is only allowed for kms v1

**Review by liggitt — COMMENTED** [review]
Talked offline with @aramase and @enj

Before beta, I think kms v2 should have *some* capability for coasting with an offline kms grpc (e.g. a minimum cache size), but can be added in a follow-up.

During beta, we could evaluate whether that minimum cache size needed to be user tunable

**Comment by aramase** [comment]
@liggitt thanks for the review. Updated the PR!

**Comment by liggitt** [comment]
API bits lgtm, CI is having unrelated issues due to python2/python3 migrations at the moment...

**Comment by liggitt** [comment]
/approve
for API bits

/lgtm
per @enj's earlier review

**Pull request merged**
